src/ops/checked.rs

```diff
@@ -1,18 +1,42 @@
-use core::ops::{Add, Div, Mul, Rem, Shl, Shr, Sub};
-
 /// Performs addition that returns `None` instead of wrapping around on
 /// overflow.
-pub trait CheckedAdd: Sized + Add<Self, Output = Self> {
+pub trait CheckedAdd<Rhs = Self>: Sized {
+    type Output;
     /// Adds two numbers, checking for overflow. If overflow happens, `None` is
     /// returned.
-    fn checked_add(&self, v: &Self) -> Option<Self>;
+    fn checked_add(self, v: Rhs) -> Option<Self::Output>;
 }
 
 macro_rules! checked_impl {
     ($trait_name:ident, $method:ident, $t:ty) => {
-        impl $trait_name for $t {
+        impl $trait_name<$t> for $t {
+            type Output = $t;
+            #[inline]
+            fn $method(self, v: $t) -> Option<$t> {
+                <$t>::$method(self, v)
+            }
+        }
+
+        impl $trait_name<&$t> for $t {
+            type Output = $t;
+            #[inline]
+            fn $method(self, v: &$t) -> Option<$t> {
+                <$t>::$method(self, *v)
```

Comment:
doesn't this force the `$t` to be `Copy`?

Reply:
I guess that assumption was already present in the existing code. I see it used below

Reply:
yeah, it's fine to assume we only care about copyable types I think